compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,390 @@
+          // See: SymDenotation.opaqueToBounds
+          case cls: ClassInfo if sym.is(Flags.Opaque) =>
+            cls.classSymbol.asClass.givenSelfType match
+              case rt: RefinedType =>
+                refinementSymtab((rt, sym.name)) = sym
+              case _ => ()
+
+          case cls: ClassInfo if (cls.cls.name == tpnme.REFINE_CLASS) =>
+            enterRefined(sym.owner.owner.info)
+
+          // type x = Person { refinement }
+          case tb: TypeBounds =>
+            // tb = TypeBounds(
+            //   lo = RefinedType(...)
+            //   hi = RefinedType(...)
+            // )
+            enterRefined(tb.lo)
+            enterRefined(tb.hi)
+
+          // def s(x: Int): { refinement } = ...
+          case expr: ExprType =>
+            enterRefined(expr.resType)
+          case m: LambdaType =>
+            enterRefined(m.resType)
+          case _ => ()
+        }
+      enterParamRef(sym.owner.info)
+      enterRefined(sym.owner.info)
+
+      def loop(tpe: Type): s.Signature = tpe match {
+        case mp: MethodOrPoly =>
+          def flatten(
+            t: Type,
+            paramss: List[List[Symbol]],
+            tparams: List[Symbol]
+          ): (Type, List[List[Symbol]], List[Symbol]) = t match {
+            case mt: MethodType =>
+              val syms = mt.paramNames.flatMap { paramName =>
+                paramRefSymtab.getOrErr((mt, paramName), sym)
+              }
+              flatten(mt.resType, paramss :+ syms, tparams)
+            case pt: PolyType =>
+              val syms = pt.paramNames.flatMap { paramName =>
+                paramRefSymtab.getOrErr((pt, paramName), sym)
+              }
+              // there shouldn't multiple type params
+              flatten(pt.resType, paramss, tparams ++ syms)
+            case other =>
+              (other, paramss, tparams)
+          }
+          val (resType, paramss, tparams) = flatten(mp, Nil, Nil)
+
+          val sparamss = paramss.map(_.sscope)
+          val stparams = tparams.sscopeOpt
+          s.MethodSignature(
+            stparams,
+            sparamss,
+            resType.toSemanticType(sym)
+          )
+
+        case cls: ClassInfo =>
+          val stparams = cls.cls.typeParams.sscopeOpt
+          val sparents = cls.parents.map(_.toSemanticType(sym))
+          val sself = cls.selfType.toSemanticType(sym)
+          val decls = cls.decls.toList.sscopeOpt
+          s.ClassSignature(stparams, sparents, sself, decls)
+
+        case TypeBounds(lo, hi) =>
+          // for `type X[T] = T` is equivalent to `[T] =>> T`
+          def tparams(tpe: Type): (Type, List[Symbol]) = tpe match {
+            case lambda: HKTypeLambda =>
+              val paramSyms = lambda.paramNames.zip(lambda.paramInfos).flatMap { (paramName, bounds) =>
+                // def x[T[_]] = ???
+                if paramName.isWildcard then
+                  val wildcardSym = newSymbol(NoSymbol, tpnme.WILDCARD, Flags.EmptyFlags, bounds)
```

Comment:
in the scala 2 implementation, a wildcard type parameter of a type parameter is a global symbol,
e.g.
```scala
class C {
  def typeLambda1[M[_]] = ???
}
```
will generate

```
Symbols:
_empty_/C# => class C extends AnyRef { +2 decls }
_empty_/C#`<init>`(). => primary ctor <init>()
_empty_/C#typeLambda1(). => method typeLambda1[M[_]]: Nothing
_empty_/C#typeLambda1().[M] => typeparam M[_]
_empty_/C#typeLambda1().[M][_] => typeparam _
```

Reply:
Made it global 👍  (but it doesn't appear in the Symbols section) https://github.com/lampepfl/dotty/pull/12885/commits/95af95c262ffaa31b3e375136aaf80cd13677ed6
Besides, if the scope contains a fake symbol (e.g. wildcard type symbol or fake symbol that represents type param), made it hard-link, because those fake symbols never appear in Symbol section, and if we symlink those fake symbols, we always alias to no symbol.
https://github.com/lampepfl/dotty/pull/12885/commits/49250915e0d28f3450f03d16a24c14577bc17562

I think hard-linking symbols are ok because
- the only drawback of hard-link is an increased size of SemanticDB by duplicating symbols both in the Symbols section and hard-linked symbols.
- However, since the fake symbol doesn't exist in Symbol section.

Reply:
I think we should still try to replicate what Scala 2 does for what it considers global symbols, so all global symbols should appear in the symbols section

Reply:
Thank you for the comment, maybe we should discuss the design in another issues but here's my opinion:

In terms of wildcard symbols, maybe we should change them to local symbols both in Scala2 and Scala3: because

- Global Symbols are the symbols that can be referenced outside the document where the definition is located
  - However, wildcard symbols are never referred by anything.
  - Therefore, I think either (local or global) is fine for wildcard symbols
- If we should replicate the SemanticDB from Scala3 to Scala2 as much as possible, I think it would be better to make wildcard symbols in Scala2 to local
  - because they are never referred by others and it's easier to implement in Scala3

Reply:
discussion continues here ... https://github.com/lampepfl/dotty/pull/12885#discussion_r678361643